File: Abstracta.py
from abc import ABC, abstractclassmethod
import abc

class Player(ABC):
    def __init__(self):
        self.isPlaying = False
        self.isOpen = True
        self.value = 0

    #metodo abtracto 
    @abstractclassmethod
    def open(filepath):
        pass

    def play(self):
        if self.isOpen:
            self.isPlaying = True
            print("The audiofile is playing.")

    def stop(self):
        if self.isPlaying: 
            self.isPlaying = False
            print("The audiofile is stopped.")

    def setvolume(self, value):
        self.value = value 
        return self.value

#*******************************************************************                                  

class VLC(Player):

    def __init__(self):
        self.pitch = 0
        self.isOpen = True
        self.filePath = ""
        print("The VLC constructor was invoked.")

    # Metodo abtracto que debo implementar en todas las clases derivadas 
    def open(self, filepath):
        self.filePath = filepath
        if self.isOpen:
            print("The audiofile: " + self.filePath + " is open.")

    def setPitch(self, value):   
        self.pitch = value
        print("The pitch value is: " + self.pitch)

#********************************************************************

class Winamp(Player):
    def __init__(self):
        self.filePath = ""
        self.isOpen = True
        print("The Winamp constructor was invoked.")
    
    # Metodo abtracto que debo implementar en todas las clases derivadas  
    def open(self, filepath):
        self.filePath = filepath
        if self.isOpen:
            print("The audiofile: " + self.filePath + " is open.")

vlcPlayer = VLC()
vlcPlayer.open("./resources/orchestral.ogg")
vlcPlayer.play()
print(" ")
print(" ")

winampPlayer = Winamp()
winampPlayer.open("./resources/orchestral.ogg")
winampPlayer.play()

File: Abstracta.java
//Player abstract class
public abstract class Player
{
    protected Boolean isOpen = false;
    protected Boolean isPlaying = false;
    protected float volume;

    public Player()
    {
        System.out.println("The Player constructor was invoked.");
        //método abstracto o virtual puro:
        public abstract void open (String filePath);
    }

    public void play()
    {
        if(isOpen) 
            this.isPlaying = true;
            System.out.println("The audiofile is playing.");
    }
    public void stop()
    {
        if(isPlaying) 
            this.isPlaying = false;
            System.out.println("The audiofile is stopped.");
    }

    public void setVolume(float value)
    {
        this.volume = value;
        System.out.println("The volume value is: " + volume);
    }
    
};
//VLC concrete class

public class VLC  extends Player
{
    private float pitch;

    public VLC()
    {
        System.out.println("The VLC constructor was invoked.");
    }
    @Override
    public void open(String filePath)
    {
       
        this.isOpen = true;
        System.out.println("The audiofile: " + filePath + " is open.");
    }

    public void setPitch(float value)
    {
        this.pitch = value;
        System.out.println("The pitch value is: " + this.pitch);
    }


};
//Winamp concrete class
public class Winamp  extends Player
{
    public Winamp()
    {
        System.out.println("The Winamp constructor was invoked.");
    }
    @Override
    public void open(String filePath)
    {
        this.isOpen = true;
        System.out.println("The audiofile: " + filePath + " is open.");
    }
};

public static void main(String[] args)
{

    VLC vlcPlayer = new VLC();
    vlcPlayer.open("./resources/orchestral.ogg");
    vlcPlayer.play();
    vlcPlayer.setVolume(13);

    Winamp winampPlayer = new Winamp();
    winampPlayer.open("./resources/orchestral.ogg");
    winampPlayer.play();
    winampPlayer.setVolume(13);
}
